File: javascript/micm/solver_result.js
/**
 * Module for solver result types from MICM.
 * 
 * This module provides:
 * - SolverState: Enum representing the final state of the solver
 * - SolverStats: Statistics from a solver run
 * - SolverResult: Combined result containing both state and statistics
 */

/**
 * Enum representing the state of the solver after execution
 */
const SolverState = {
  NotYetCalled: 0,
  Running: 1,
  Converged: 2,
  ConvergenceExceededMaxSteps: 3,
  StepSizeTooSmall: 4,
  RepeatedlySingularMatrix: 5,
  NaNDetected: 6,
  InfDetected: 7,
  AcceptingUnconvergedIntegration: 8,
};

/**
 * Class representing solver statistics
 */
class SolverStats {
  constructor({
    function_calls = 0,
    jacobian_updates = 0,
    number_of_steps = 0,
    accepted = 0,
    rejected = 0,
    decompositions = 0,
    solves = 0,
    final_time = 0.0,
  } = {}) {
    this.function_calls = function_calls;
    this.jacobian_updates = jacobian_updates;
    this.number_of_steps = number_of_steps;
    this.accepted = accepted;
    this.rejected = rejected;
    this.decompositions = decompositions;
    this.solves = solves;
    this.final_time = final_time;
  }

  toString() {
    return `SolverStats{function_calls: ${this.function_calls}, jacobian_updates: ${this.jacobian_updates}, number_of_steps: ${this.number_of_steps}, accepted: ${this.accepted}, rejected: ${this.rejected}, decompositions: ${this.decompositions}, solves: ${this.solves}, final_time: ${this.final_time}}`;
  }
}

/**
 * Class representing the combined solver result
 */
class SolverResult {
  /**
   * @param {number} state - The solver state (from SolverState enum)
   * @param {SolverStats} stats - The solver statistics
   */
  constructor(state, stats) {
    this.state = state;
    this.stats = stats;
  }

  toString() {
    const stateName = Object.keys(SolverState).find(
      (key) => SolverState[key] === this.state
    );
    return `SolverResult{state: ${stateName}, stats: ${this.stats}}`;
  }
}

module.exports = { SolverState, SolverStats, SolverResult };

File: julia/src/micm/solver_result.jl
# Copyright (C) 2023-2026 University Corporation for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0

"""
    SolverState

Enum representing the final state of the solver after a solve call.
"""
@enum SolverState begin
    NotYetCalled = 0
    Running = 1
    Converged = 2
    ConvergenceExceededMaxSteps = 3
    StepSizeTooSmall = 4
    RepeatedlySingularMatrix = 5
    NaNDetected = 6
    InfDetected = 7
    AcceptingUnconvergedIntegration = 8
end

"""
    SolverStats

Statistics from a solver run.
"""
struct SolverStats
    function_calls::Int
    jacobian_updates::Int
    number_of_steps::Int
    accepted::Int
    rejected::Int
    decompositions::Int
    solves::Int
    final_time::Float64
end

"""
    SolverResult

Combined result containing both solver state and statistics.
"""
struct SolverResult
    state::SolverState
    stats::SolverStats
end

function SolverResult(cpp_result::CppSolverResult)
    cpp_stats = get_solver_stats(cpp_result)
    stats = SolverStats(
        get_function_calls(cpp_stats),
        get_jacobian_updates(cpp_stats),
        get_number_of_steps(cpp_stats),
        get_accepted(cpp_stats),
        get_rejected(cpp_stats),
        get_decompositions(cpp_stats),
        get_solves(cpp_stats),
        get_final_time(cpp_stats),
    )
    state = SolverState(get_solver_state(cpp_result))
    return SolverResult(state, stats)
end

function Base.show(io::IO, r::SolverResult)
    print(
        io,
        "SolverResult(state=$(r.state), steps=$(r.stats.number_of_steps), final_time=$(r.stats.final_time))",
    )
end
